pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -308,11 +334,49 @@ func (ctrl *PersistentVolumeController) shouldDelayBinding(claim *v1.PersistentV
 	if ctrl.isDelayBindingProvisioning(claim) {
 		return false, nil
 	}
-
 	// If claim is in delay binding mode.
 	return pvutil.IsDelayBindingMode(claim, ctrl.classLister)
 }
 
+// Unit test [12-2] [12-4]
+func (ctrl *PersistentVolumeController) createOperationTimestampCacheKey(opName, claimKey string) string {
+	return claimKey + "-" + opName
+}
+
+func (ctrl *PersistentVolumeController) recordCSIMetric(opName, claimKey string, err error) {
+	operationTsKey := ctrl.createOperationTimestampCacheKey(opName, claimKey)
+	obj, exists := ctrl.operationTimestamps.Get(operationTsKey)
+	if !exists {
+		// in this case, an entry has not been created yet by either provisionClaim or deleteVolume
+		// there are two scenarios:
+		// 1. volume already existed, thus syncUnboundClaim will *NOT* call provisionClaim
+		//    rather just bind it to a PVC, in this case, as the latency will not be reported
+		//    as it does not make much sense (will be 0). For this situation, a possible improvement
+		//    could be instead using claim's creationTimestamp in metadata, however will bring
+		//    in-consistency. Ignoring here for now
+		// 2. the plugin.IsMigratedToCSI() == false, thus provisionClaim will *NOT* create
+		//    an entry for metric recording, in this case, return directly
+		//
+		return
+	}
+	operationTs, ok := obj.(operationTimestamp)
+	if !ok {
+		klog.V(2).Infof("Cannot convert object from operationTimestamps cache to operationTimeStamp for volume claim key[%s], operation[%s]: %#v", claimKey, opName, obj)
+		return
+	}
+	// in case of error, report count instead of latency, handled by RecordVolumeOperationMetric
+	if err != nil {
+		// no timeTaken is needed
+		metrics.RecordVolumeOperationMetric(operationTs.provisionerName, opName, 0.0, err)
+	} else {
+		// this is safe as the idempotency nature of storage operations
+		timeTaken := time.Since(operationTs.startTs).Seconds()
+		metrics.RecordVolumeOperationMetric(operationTs.provisionerName, opName, timeTaken, nil)
+		// remove
+		ctrl.operationTimestamps.Delete(operationTsKey)
```

Comment:
We need to delete the key regardless of error condition or not.

Reply:
Can we not cleanup the metric when the PV object gets deleted (via informer event)?

Reply:
The problem is, an entry is added to threadsafe cache store when PVC is created. PVC may or may not be provisioned but cache still gets an entry.

Reply:
My current thoughts for managing the lifecycle of the cache:
1. after a PVC is bound to PV, report provision latency and delete the cache
2. when a PVC is deleted from system:
    a. if PVC is bound, report deletion latency if needed and delete the cache (there should NOT be any entry in the cache for provisioning in this case)
   b. if PVC is not bound, delete the cache if exists (This is currently under implementation along with the deletion latency metric)
3. when a PV is deleted from system and the corresponding PVC has not been bound to it yet, remove the cache without reporting
This might have the problem @gnufied mentioned, if a PVC has been created but never get provisioned. My understanding is such a PVC will exist forever in the system until get deleted, and the controller will keep on retrying. I will think its rather legit?

Reply:
cc @jsafrane

Reply:
thats how its implemented currently 👍